nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java

```diff
@@ -0,0 +1,150 @@
+ * the consequent is "!null" is supported. The handler checks and warns under the conditions of the
+ * antecedent if the consequent is "!null" and there is a return statement with "nullable" or "null"
+ * expression.
+ */
+public class ContractCheckHandler extends BaseNoOpHandler {
+
+  private Set<TreePath> supportedMethods = new HashSet<>();
+
+  @Override
+  public void onMatchMethod(
+      NullAway analysis, MethodTree tree, VisitorState state, Symbol.MethodSymbol methodSymbol) {
+    Symbol.MethodSymbol callee = ASTHelpers.getSymbol(tree);
+    Preconditions.checkNotNull(callee);
+    // Check to see if this method has an @Contract annotation
+    String contractString = getContractFromAnnotation(callee);
+
+    if (contractString != null) {
+      // Found a contract, lets parse it.
+      String[] clauses = contractString.split(";");
+      if (clauses.length != 1) {
+        return;
+      }
+
+      String clause = clauses[0];
+
+      String[] parts = clause.split("->");
+      if (parts.length != 2) {
+        reportMatch(
+            tree,
+            "Invalid @Contract annotation detected for method "
+                + callee
+                + ". It contains the following uparseable clause: "
+                + clause
+                + "(see https://www.jetbrains.com/help/idea/contract-annotations.html).",
+            analysis,
+            state);
+      }
+      String[] antecedent = parts[0].split(",");
+      String consequent = parts[1].trim();
+
+      if (antecedent.length != tree.getParameters().size()) {
+        reportMatch(
+            tree,
+            "Invalid @Contract annotation detected for method "
+                + callee
+                + ". It contains the following uparseable clause: "
+                + clause
+                + " (incorrect number of arguments in the clause's antecedent ["
+                + antecedent.length
+                + "], should be the same as the number of "
+                + "arguments in for the method ["
+                + tree.getParameters().size()
+                + "]).",
+            analysis,
+            state);
+      }
+
+      boolean supported = true;
+
+      for (int i = 0; i < antecedent.length; ++i) {
+        String valueConstraint = antecedent[i].trim();
+        if (!(valueConstraint.equals("_")
+            || valueConstraint.equals("!null")
+            || valueConstraint.equals("null"))) {
+          supported = false;
+        }
+      }
+
+      if (!consequent.equals("!null")) {
+        supported = false;
+      }
+
+      if (!supported) {
+        return;
+      }
+
+      supportedMethods.add(state.getPath());
+    }
+  }
+
+  @Override
+  public void onMatchReturn(NullAway analysis, ReturnTree tree, VisitorState state) {
+    final TreePath enclosingMethod =
+        NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(state.getPath());
+    if (!supportedMethods.contains(enclosingMethod)) {
```

Comment:
This `supportedMethods` set is a bit messy and also may be a small memory leak (it never gets cleared).  Maybe instead of doing the check in `onMatchReturn`, we should do a separate traversal inside `onMatchMethod` that finds and checks return statements and reports errors from there?  It is extra traversal, but only for `@Contract` methods.  It might be more efficient than having to look up the enclosing method for every `return` statement.

Reply:
I am not sure, how this can be done. I might be able to find `ReturnTree` from `MethodTree`. but, not sure how to get the correct `visitorState`.

Reply:
Removed the memory leak problem. 

For the alternate method, I am not sure, how this can be done. we might be able to find `ReturnTree` from `MethodTree`. but, not sure how to get the correct `visitorState`.

Reply:
We can create an appropriate `VisitorState` using the `withPath` method:

https://github.com/google/error-prone/blob/1e2c8ac094d6d8b8c968f722d723800c5a79fb20/check_api/src/main/java/com/google/errorprone/VisitorState.java#L219-L221

I think we should change to doing our own traversal using a `TreePathScanner` inside `onMatchMethod` instead of using `onMatchReturn`.  This way we won't have the overhead of finding the containing method for every `return` statement in the program.

Reply:
Agree with @msridhar here.

Reply:
Done